# Need samtools and bowtie2 in path to work.
# Execute with command: snakemake ./data/mapping/4_mapped.bam

# Inspo: https://github.com/EnvGen/snakemake-workflows/blob/master/bio/ngs/rules/mapping/bowtie2.rules

# Need to create config file and make script more general.

import re
configfile: "config.json"


def mapping_bowtie2_command():
    if len(config["bowtie2"]["reads"])==2:
        return " -1 "+config["bowtie2"]["reads"][0]+" -2 "+config["bowtie2"]["reads"][1]+" "
    elif len(config["bowtie2"]["reads"])==1:
        return " -U "+config["bowtie2"]["reads"][0]+" "
    else:
        raise(Exception("Reads should be paired or single ended."))


rule all:
    input:
        "data/mapping/4_mapped.bam"

# path to index file without extension (removed .fa or .fasta)
index_basename="data/index/"+re.split('/|\.',config["bowtie2"]["reference"])[-2]

rule bowtie2_index:
    input:
        config["bowtie2"]["reference"]
    output:
        expand(index_basename+".{index}.bt2", index=range(1,5)),
        expand(index_basename+".rev.{index}.bt2", index=range(1,3))
    run:
        shell("bowtie2-build {input} "+index_basename)

rule bowtie2_map:
    input:
        # Indexed files
        expand(index_basename+".{index}.bt2", index=range(1,5)),
        expand(index_basename+".rev.{index}.bt2", index=range(1,3)),
        # Read files
        config["bowtie2"]["reads"]
    output:
        "data/mapping/4_mapped.bam"
    params:
        ref_idx_base=index_basename
    threads: 4 # for my laptop
    #shell: # need to add more different params here that changes according to config-file.
    #    "bowtie2 -x {params.ref_idx_base} -1 {input.left} -2 {input.right} | samtools view -bS - > {output}"
    run: # need to add more different params here that changes according to config-file.
        shell("bowtie2 \
        -x {params.ref_idx_base} \
        "+mapping_bowtie2_command()+"\
        -p {threads} \
        | samtools view -bS - > {output}")
